lib/ukstore/store.c

```diff
@@ -29,3 +29,1186 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  * POSSIBILITY OF SUCH DAMAGE.
  */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <uk/arch/limits.h>
+#include <uk/arch/types.h>
+#include <uk/store.h>
+
+struct uk_store_folder {
+	/* List for the dynamic folders only */
+	struct uk_list_head folder_head;
+
+	/* List for the entries in the folder */
+	struct uk_list_head entry_head;
+
+	/* Allocator to be used in allocations */
+	struct uk_alloc *a;
+
+	/* The folder name */
+	char *name;
+} __align8;
+
+struct uk_store_folder_entry {
+	/* Saved entry */
+	struct uk_store_entry entry;
+
+	/* Folder reference */
+	struct uk_store_folder *folder;
+
+	/* Allocator to be used in allocations */
+	struct uk_alloc *a;
+
+	/* Cleanup function */
+	uk_store_entry_freed_func_t clean;
+
+	/* The linked list head */
+	struct uk_list_head list_head;
+
+	/* Refcount */
+	__atomic refcount;
+} __align8;
+
+#define UK_STORE_NR_MAX_LEN 32
```

Comment:
I would suggest a different name. It wasn't clear to me what these definitions mean until I saw the code using them. Additionally, they are also internal only, so you do not need the `UK_STORE` prefix.
What would you think of:
```c
#define _U8_STRLEN (3 + 1) /* biggest num (127) has 3 digits plus terminating '\0' */
#define _S8_STRLEN (1 + _U8_STRLEN) /* space for potential sign in front */
#define _U16_STRLEN (5 + 1)
#define _S16_STRLEN (1 + _U16_STRLEN)
#define _U32_STRLEN (10 + 1)
#define _S32_STRLEN (1 + _U32_STRLEN)
#define _U64_STRLEN (19 + 1)
#define _S64_STRLEN (1 + _U64_STRLEN)
#define _UPTR_STRLEN (2 + 16 + 1) /* this is for uptr (64bit) in hex with space for leading `0x` prefix */
```